Server/ConcordiaCurriculumManager/Models/Curriculum/CourseGroupings/CourseGrouping.cs

```diff
@@ -52,6 +53,51 @@ private static CourseGrouping CreateCourseGroupingFromDTO(Guid id,CourseGrouping
             Courses = new List<Course>()
         };
     }
+
+    public bool IsCourseGroupingStateFinalized() => State == CourseGroupingStateEnum.Accepted || State == CourseGroupingStateEnum.Deleted;
+
+    public void MarkAsPublished()
+    {
+        Published = true;
+
+        VerifyCourseGroupingIsValidOrThrow();
+    }
+
+    public void MarkAsUnpublished()
+    {
+        Published = false;
+
+        VerifyCourseGroupingIsValidOrThrow();
+    }
+
+    private void VerifyCourseGroupingIsValidOrThrow()
+    {
+        if (Published)
+        {
+            if (IsCourseGroupingStateFinalized() && Version != null)
+            {
+                foreach (var courseGrouping in SubGroupings)
```

Comment:
You try and use the SubGroupings and Courses arrays here, but they wouldn't actually be loaded by the repository. In the service layer, there's already a `QueryRelatedCourseGroupingData` method I wrote that you can call to load in all the child groupings and courses that you're checking here.